Geometry_Engine/Query/GeometryHash.cs

```diff
@@ -0,0 +1,427 @@
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.Engine.Base;
+using BH.oM.Base;
+using BH.oM.Base.Attributes;
+using BH.oM.Geometry;
+using BH.oM.Geometry.CoordinateSystem;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Drawing.Drawing2D;
+using System.Linq;
+using System.Security.Cryptography.Xml;
+
+namespace BH.Engine.Geometry
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /****               Public Methods              ****/
+        /***************************************************/
+
+        [Description("Returns a signature of the input geometry, useful for distance-based comparisons and diffing." +
+            "\nThe hash is computed as an array representing the coordinate of significant points taken on the geometry." +
+            "\nThe number of points is reduced to the minimum essential to determine uniquely any geometry." +
+            "\nAdditionally, the resulting points are transformed based on the source geometry type, to remove or minimize collisions." +
+            "\n(Any transformation so performed is translational only, in order to support geometrical tolerance, i.e. numerical distance, when comparing GeometryHashes downstream).")]
+        [Output("geomHash", "Array of numbers representing a unique signature of the input geometry.")]
+        public static double[] IGeometryHash(this IGeometry igeometry)
+        {
+            return GeometryHash(igeometry as dynamic, 0);
+        }
+
+        /***************************************************/
+        /****              Private Methods              ****/
+        /***************************************************/
+
+
+        /***************************************************/
+        /****  Curves                                   ****/
+        /***************************************************/
+
+        [Description("The geometry hash of a Curve is obtained by first retrieving any Sub-part of the curve, if present." +
+            "The ISubParts() methods is able to return the 'primitive' curves that a curve is composed of. " +
+            "The GeometryHashes are then calculated for the individual parts and concatenated.")]
+        private static double[] GeometryHash(this ICurve curve, double translationFactor)
+        {
+            IEnumerable<ICurve> subParts = curve.ISubParts();
+
+            bool isMultipleCurves = subParts.Count() != 1;
+
+            return subParts.SelectMany(c => (double[])GeometryHash(c as dynamic, translationFactor, isMultipleCurves)).ToArray();
```

Comment:
This we want the endpoint of the last curve included - at least for open curves, but probably fine to include for all. Would change to something like:
```suggestion
            List<ICurve> subParts = curve.ISubParts().ToList();

            List<double> hashes = new List<double>();

            //Add hash ignoring endpoint for all but last curve
            for (int i = 0; i < subParts.Count - 1; i++)
            {
                hashes.Add(GeometryHash(subParts[i] as dynamic, translationFactor, true));
            }
            //Include endpoint for hasing for last curve
            hashes.Add(GeometryHash(subParts.Last() as dynamic, translationFactor, false));

            return hashes.ToArray();
```

Could add a check for IsClosed for whether the endpoint of the last curve should be included or not, but then we start running in to various tolerance discussions that I think we want to avoid, so probably better to just always include it as above.